autogen/src/header.rs

```diff
@@ -43,9 +44,55 @@ fn bit_enum_attribute() -> TokenStream {
     }
 }
 
-fn from_primitive_impl(from_prim: &[TokenStream], kind: &proc_macro2::Ident) -> TokenStream {
+fn generate_enum(
+    enum_name: &proc_macro2::Ident,
+    variants: &[(u32, proc_macro2::Ident)],
+    comment: String,
+) -> TokenStream {
+    let mut variants = variants.to_vec();
+    variants.sort_by_key(|&(number, _)| number);
+    let enumerants = variants
+        .iter()
+        .map(|(number, name)| quote! { #name = #number });
+
+    let mut number_runs = vec![];
+    number_runs.push(variants[0].0..=variants[0].0);
+    for &(number, _) in variants.iter().skip(1) {
+        let last_run = number_runs.last_mut().unwrap();
+        if number == *last_run.end() + 1 {
+            *last_run = *last_run.start()..=*last_run.end() + 1;
```

Comment:
Just:
```suggestion
            *last_run = *last_run.start()..=number;
```
?

I wonder if there's a more concise way of writing this though.

Reply:
Unfortunately there isn't. `RangeInclusive` doesn't have any setters.

Reply:
Yeah, and keeping a min and max in a `u32` separately, updating the max bounds, and adding it whenever the next number is not contiguous _or we run out of numbers_ (not to mention resetting min and max) costs many more lines, too.

Reply:
Even though I'm a fan of `RangeInclusive` none of this code takes advantage of it (it's only used as `RangeInclusive` in the final generated code) so you might as well use a tuple for the min and max which you _can_ update in place?

Reply:
Like this?